### PR TITLE
Check for missing hierarchies when mapping to Roslyn projects

### LINKED ISSUE #19517: NullReferenceException at SolutionExplorer.HierarchyItemToProjectIdMap.TryGetProjectId
**Version Used**: 
VS: vsuvscore 62510.01
Roslyn: 2.03.0.61709, commit c4b1692

**Steps to Reproduce**:

1. Create a C# Desktop class library project
2. Add a C# console application project to the solution
3. Enable Lightweight solution load mode
4. Save the solution, restart VS, load the solution
5. Expand console app project node in Solution Explorer

**Expected Behavior**:
The project items are under the project node

**Actual Behavior**:
No project items under the project node

I've tracked this to _NullReferenceException_ happening at _SolutionExplorer.HierarchyItemToProjectIdMap.TryGetProjectId_, see the call stack below.

In line src\VisualStudio\Core\SolutionExplorerShim\HierarchyItemToProjectIdMap.cs (56)
__p.Hierarchy__ is null:
```csharp
if (p.Hierarchy.TryGetCanonicalName((uint)VSConstants.VSITEMID.Root, out string projectCanonicalName)
```

**Call Stack**
```
>	Microsoft.VisualStudio.LanguageServices.SolutionExplorer.dll!Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer.HierarchyItemToProjectIdMap.TryGetProjectId.AnonymousMethod__0(Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.AbstractProject p) Line 56	C#
 	System.Core.dll!System.Linq.Enumerable.WhereArrayIterator<Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.AbstractProject>.MoveNext() Line 189	C#
 	System.Core.dll!System.Linq.Enumerable.SingleOrDefault<Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.AbstractProject>(System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.AbstractProject> source) Line 1084	C#
 	Microsoft.VisualStudio.LanguageServices.SolutionExplorer.dll!Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer.HierarchyItemToProjectIdMap.TryGetProjectId(Microsoft.VisualStudio.Shell.IVsHierarchyItem hierarchyItem, string targetFrameworkMoniker, out Microsoft.CodeAnalysis.ProjectId projectId) Line 74	C#
 	Microsoft.VisualStudio.LanguageServices.SolutionExplorer.dll!Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer.AnalyzersFolderItemProvider.CreateCollectionSourceCore(Microsoft.VisualStudio.Shell.IVsHierarchyItem parentItem, Microsoft.VisualStudio.Shell.IVsHierarchyItem item) Line 70	C#
 	Microsoft.VisualStudio.LanguageServices.SolutionExplorer.dll!Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer.AnalyzersFolderItemProvider.CreateCollectionSource(Microsoft.VisualStudio.Shell.IVsHierarchyItem item, string relationshipName) Line 63	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.AttachedCollectionSourceProvider<Microsoft.VisualStudio.Shell.IVsHierarchyItem>.Microsoft.VisualStudio.Shell.IAttachedCollectionSourceProvider.CreateCollectionSource(object item, string relationshipName) Line 21	C#
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.AttachedCollectionService.AggregateItemsSource.AnonymousMethod__0(System.Lazy<Microsoft.VisualStudio.Shell.IAttachedCollectionSourceProvider, Microsoft.VisualStudio.Utilities.IOrderable> provider) Line 140	C#
 	System.Core.dll!System.Linq.Enumerable.WhereSelectListIterator<System.Lazy<Microsoft.VisualStudio.Shell.IAttachedCollectionSourceProvider, Microsoft.VisualStudio.Utilities.IOrderable>, Microsoft.VisualStudio.Shell.IAttachedCollectionSource>.MoveNext() Line 372	C#
 	System.Core.dll!System.Linq.Enumerable.WhereEnumerableIterator<Microsoft.VisualStudio.Shell.IAttachedCollectionSource>.MoveNext() Line 147	C#
 	mscorlib.dll!System.Collections.Generic.List<Microsoft.VisualStudio.Shell.IAttachedCollectionSource>.List(System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.Shell.IAttachedCollectionSource> collection)	Unknown
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.AggregateAttachedCollectionSource.AggregateAttachedCollectionSource(object sourceItem, System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.Shell.IAttachedCollectionSource> sources) Line 27	C#
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.AttachedCollectionService.AggregateItemsSource(object item, string relationshipName) Line 146	C#
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.AttachedCollectionService.GetOrCreateCollectionSource(object item, string relationshipName) Line 116	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.PivotNavigationEntry.ConvertItemToCollectionSource(object item) Line 83	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.ConvertItemToCollectionSource(object item) Line 124	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.ItemCollectionAdapter.AdaptItem(object item) Line 3310	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.CollectionAdapter<object, object>.Initialize(System.Collections.IEnumerable source) Line 87	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.ItemCollectionAdapter.ItemCollectionAdapter(System.Collections.IEnumerable source, Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView treeView) Line 3295	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.ItemCollectionConverter.Convert(object value, System.Type targetType, object parameter, System.Globalization.CultureInfo culture) Line 3269	C#
 	PresentationFramework.dll!System.Windows.Data.BindingExpression.TransferValue(object newValue, bool isASubPropertyChange)	Unknown
 	PresentationFramework.dll!System.Windows.Data.BindingExpression.Activate(object item)	Unknown
 	PresentationFramework.dll!System.Windows.Data.BindingExpression.AttachToContext(System.Windows.Data.BindingExpression.AttachAttempt attempt)	Unknown
 	PresentationFramework.dll!System.Windows.Data.BindingExpression.AttachOverride(System.Windows.DependencyObject target, System.Windows.DependencyProperty dp)	Unknown
 	PresentationFramework.dll!System.Windows.Data.BindingExpressionBase.OnAttach(System.Windows.DependencyObject d, System.Windows.DependencyProperty dp)	Unknown
 	WindowsBase.dll!System.Windows.DependencyObject.SetValueCommon(System.Windows.DependencyProperty dp, object value, System.Windows.PropertyMetadata metadata, bool coerceWithDeferredReference, bool coerceWithCurrentValue, System.Windows.OperationType operationType, bool isInternal)	Unknown
 	PresentationFramework.dll!System.Windows.Data.BindingOperations.SetBinding(System.Windows.DependencyObject target, System.Windows.DependencyProperty dp, System.Windows.Data.BindingBase binding)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.TreeNode.RealizeChildNodesCore() Line 1768	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.TreeNode.ExpandInternal() Line 1877	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.TreeNode.OnIsExpandedChanged(System.Windows.DependencyObject o, System.Windows.DependencyPropertyChangedEventArgs e) Line 1960	C#
 	WindowsBase.dll!System.Windows.DependencyObject.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs e)	Unknown
 	WindowsBase.dll!System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs args)	Unknown
 	WindowsBase.dll!System.Windows.DependencyObject.UpdateEffectiveValue(System.Windows.EntryIndex entryIndex, System.Windows.DependencyProperty dp, System.Windows.PropertyMetadata metadata, System.Windows.EffectiveValueEntry oldEntry, ref System.Windows.EffectiveValueEntry newEntry, bool coerceWithDeferredReference, bool coerceWithCurrentValue, System.Windows.OperationType operationType)	Unknown
 	WindowsBase.dll!System.Windows.DependencyObject.SetValueCommon(System.Windows.DependencyProperty dp, object value, System.Windows.PropertyMetadata metadata, bool coerceWithDeferredReference, bool coerceWithCurrentValue, System.Windows.OperationType operationType, bool isInternal)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeView.TreeNode.IsExpanded.set(bool value) Line 1529	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.HierarchyNavigatorPaneBase<System.__Canon>.ExpandItem.AnonymousMethod__0() Line 750	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.ErrorHandler.CallWithCOMConvention(System.Func<int> method, bool reportError, bool setShellErrorInfo) Line 165	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.HierarchyNavigatorPaneBase<Microsoft.VisualStudio.PlatformUI.SolutionPivotNavigator>.ExpandItem(Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchy pUIH, uint itemid, Microsoft.VisualStudio.Shell.Interop.EXPANDFLAGS grf) Line 651	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.SolutionNavigatorPaneMarshaler.Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchyWindow.ExpandItem.AnonymousMethod__0() Line 108	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.ThreadHelper.Invoke<int>(System.Func<int> method) Line 362	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.SolutionNavigatorPaneMarshaler.Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchyWindow.ExpandItem(Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchy pUIH, uint itemid, Microsoft.VisualStudio.Shell.Interop.EXPANDFLAGS expf) Line 108	C#
 	[Native to Managed Transition]	
 	msenv.dll!ExpandItemInSolutionExplorer(IUnknown * pUnkHier, unsigned long itemid, __EXPANDFLAGS fExpand) Line 612	C++
 	msenv.dll!CProject::RefreshSolutionExplorerState() Line 3409	C++
 	msenv.dll!CProject::FinalizeProvisionedProjectAsync::__l2::<lambda>(IVsTask *) Line 2076	C++
 	msenv.dll!std::_Func_class<long,IVsTask *,unsigned long,IVsTask * *,tagVARIANT *>::operator()(IVsTask * <_Args_0>, unsigned long <_Args_1>, IVsTask * * <_Args_2>, tagVARIANT * <_Args_3>) Line 278	C++
 	msenv.dll!`anonymous namespace'::NativeVsTaskBody::DoWork(IVsTask * pTask, unsigned long dwCount, IVsTask * * pPredecessorTasks, tagVARIANT * pResult) Line 46	C++
 	Microsoft.VisualStudio.Shell.UI.Internal.ni.dll!52277efe()	Unknown
 	[Frames below may be incorrect and/or missing, native debugger attempting to walk managed call stack]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetCallbackForMultipleParent.AnonymousMethod__0(System.Threading.Tasks.Task<object>[] _) Line 1106	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetCallbackForSingleParent.AnonymousMethod__0(System.Threading.Tasks.Task<object> internalTask) Line 1151	C#
 	mscorlib.dll!System.Threading.Tasks.ContinuationResultTaskFromResultTask<object, object>.InnerInvoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.Tasks.TaskScheduler.TryExecuteTask(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.TaskSchedulerService.VsUIThreadBlockableTaskScheduler.PromoteTaskExecution.AnonymousMethod__0() Line 263	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.InvokableAction.InvokeMethod() Line 80	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.InvokableBase.Invoke() Line 26	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.UIThreadReentrancyScope.ExecuteOne() Line 599	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.UIThreadReentrancyScope.WaitOnTaskCompleteInternal(System.Threading.Tasks.Task task, System.Threading.CancellationToken cancel, int ms) Line 825	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalWait.AnonymousMethod__2() Line 508	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.IgnoreObjectDisposedException(System.Action action) Line 670	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalWait.AnonymousMethod__0() Line 504	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog<bool>(System.Func<bool> function) Line 850	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalWait(bool ignoreUIThreadCheck, int millisecondTimeout, bool abortOnCancel) Line 485	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.Wait() Line 300	C#
 	[Native to Managed Transition]	
 	msenv.dll!CDelayProjectLoadManager::LoadSingleProjectAsync::__l2::<lambda>(IVsTask * pParentTasks, unsigned long) Line 524	C++
...

```

### PR DESCRIPTION
**Pull request by tmeschter**
In order to shows the "Analyzers" node under the "References" node in
Solution Explorer and show the correct sets of analyzer assemblies and
diagnostics, we need to map the node for the project to the actual
Roslyn `AbstractProject` instance. This is done by matching up various
properties of the `IVsHierarchy` for the node to the same properties on
the `IVsHierarchy` returned by `AbstractProject.Hierarchy`.

This code path was initially written this `Hierarchy` property would
never return null; its value was supplied by the project system.
However, Delayed Project Load (DPL) bypasses the project system until
the user expands the project node in Solution Explorer, and as a result
we may now encounter `AbstractProjects` where `Hierarchy` returns null.
Currently if we run across one of these we'll throw a
`NullReferenceException` which has the end effect of preventing the
user from expanding the project node in Solution Explorer, and may even
result in a crash.

This commit adds in the null check. If the `Hierarchy` property is null
then it can't be the project we're looking for.

**Customer scenario**

Customer opens a solution with DPL enabled. When they expand a project node in Solution Explorer, nothing happens (the arrow indicates that the project is expanded, but nothing shows up under the project node).

Subsequently searching in Solution Explorer may lead to a crash.

**Bugs this fixes:**

Fixes #19517.

**Workarounds, if any**

Disable DPL for the solution.

**Risk**

Low; we're adding a null check that always should have been there.

**Performance impact**

Low; just the cost an additional null check when the user initially expands a project.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

This code path predates DPL, and until DPL the property in question couldn't be null. A subsequent refactoring to support the new project system made use of the property, but I didn't realize the null check was now necessary.

**How was the bug found?**

Internal customer report

### CODE REVIEW

**Comment by tmeschter** [comment]
@Pilchie @dotnet/roslyn-ide Please review.

File: src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs [hunk 1]
```diff
@@ -62,7 +62,8 @@ public bool TryGetProjectId(IVsHierarchyItem hierarchyItem, string targetFramewo
                         // project files are in the same folder--they both use the full path to the _folder_ as the canonical
                         // name. To distinguish them we also examine the "regular" name, which will necessarily be different
                         // if the two projects are in the same folder.
-                        if (p.Hierarchy.TryGetCanonicalName((uint)VSConstants.VSITEMID.Root, out string projectCanonicalName)
+                        if (p.Hierarchy != null
```
Comment (tmat) [inline src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs]: Add a comment explaining when Hierarchy is null.

Comment (tmeschter) [inline src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs]: Good idea.

Comment (tmeschter) [inline src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs]: @tmat Done.

**Comment by tmeschter** [comment]
Adding @MattGertz for approval, pending tests.

**Comment by tmeschter** [comment]
@dotnet-bot test windows_release_unit32_prtest please

**Pull request merged**